### PR TITLE
refactor: AUTH_LDAP/AUTH_OAUTH + implement role mapping

### PR DESCRIPTION
**Pull request by thesuperzapper**
This PR is a significant refactor for AUTH_LDAP and AUTH_OAUTH.

Aside adding significant unit testing, and resolving many issues, it implements a feature with `AUTH_ROLES_MAPPING` to allow mapping LDAP/OAUTH groups to FAB Roles.

For more information, please see the updated `docs/security.rst`

Resolves:
- #518 
- #956 
- #1123
- https://github.com/apache/airflow/issues/8179
- https://github.com/apache/incubator-superset/issues/3419

Superseeds these PRs:
- #770

### CODE REVIEW

**Comment by thesuperzapper** [comment]
@dpgaspar can you please review this?

**Comment by thesuperzapper** [comment]
I have accidentally introduce a behaviour change for LDAP servers which bind without a search account, I will make some changes tomorrow. (+ I need to fix that failing test)

File: docs/security.rst
Comment (CameronNemo) [inline docs/security.rst:479]: nit: indentation for this brace is off.

File: docs/security.rst
Comment (CameronNemo) [inline docs/security.rst:510]: same with these two

File: flask_appbuilder/security/manager.py
Comment (CameronNemo) [inline flask_appbuilder/security/manager.py:594]: you should not need to use `elif` here and above, `if` is fine because each if clause always ends in a return statement.

File: flask_appbuilder/security/manager.py
Comment (CameronNemo) [inline flask_appbuilder/security/manager.py:906]: likewise you can remove the else and unindent the following lines, because the if above results in a return statement

File: flask_appbuilder/tests/_test_ldapsearch.py
Comment (CameronNemo) [inline flask_appbuilder/tests/_test_ldapsearch.py:65]: I am curious what these test changes accomplish?

File: flask_appbuilder/security/manager.py
Comment (dpgaspar) [reply flask_appbuilder/security/manager.py:594]: This is asking for a refactor also, we could implement a map between the providers and the functions that implement the necessary data mapping. Out of scope for this PR

**Comment by dpgaspar** [comment]
@thesuperzapper, I expect to release FAB major 3 soon, would love to get this one in.

**Comment by thesuperzapper** [comment]
@dpgaspar yeah, I will pick this up again, got a bit busy over the last few weeks.

**Comment by jedcunningham** [comment]
@thesuperzapper, gentle bump. Also happy to help get some test coverage on this. FAB 3 has already been released, so we missed that boat.

**Comment by thesuperzapper** [comment]
@jedcunningham, would welcome some help on the testing side.

I can pretty easily address the requested changes above. 
However there remains the question raised about allowing multiple role mappings for a single CN in: `AUTH_ROLES_MAPPING`.

File: docs/security.rst
Comment (thesuperzapper) [reply docs/security.rst:566]: @dpgaspar what did you end up doing in FAB 3? Is this multi-role mapping still necessary?

**Comment by jedcunningham** [comment]
@thesuperzapper, I've actually got pretty close to complete coverage on the ldapauth side of things at this point and have addressed some of the concerns raised above. I don't have an easy way to test the oauth side out though. Let me get my stuff through my $works legal and I'll get my changes out.

**Comment by accorvin** [comment]
@thesuperzapper we’d love to make use of this functionality in our project. Is there anything we can do to help get this ready for approval faster?

**Comment by thesuperzapper** [comment]
@accorvin we are currently working on it, so expect an update in a week or so.

**Comment by Karthik-13** [comment]
@thesuperzapper we require this functionality in our project as well. Any idea on when this will be released or how we can help speed this up?

**Comment by brokenjacobs** [comment]
Anything that can be done to help?

**Comment by brokenjacobs** [comment]
> @accorvin we are currently working on it, so expect an update in a week or so.

Any status?

File: flask_appbuilder/security/manager.py
Comment (annatve) [inline flask_appbuilder/security/manager.py:218]: add

**Comment by thesuperzapper** [comment]
I am very close to getting this done, I am just taking the time to also fix some security issues.

File: flask_appbuilder/security/manager.py
Comment (megakoresh) [inline flask_appbuilder/security/manager.py:579]: This dict is missing the groups claim returned from azure if you configure the enterprise app to return group information. The key name is `groups`, and it's configurable in azure, but I think it's sensible to just require the default - 99% of cases won't change that.

So just adding

```python
"groups": me.get("groups", [])
```` 

to the dict below would work. But documentation should also be updated to instruct users to configure the group claim to be sent with the token for the app, because it's optional, so not sent by default (that's why I had the `[]` there). @thesuperzapper can you add that pls?

**Comment by thesuperzapper** [comment]
I know this is taking a while, but I still have some unit tests to write for OAUTH

**Comment by thesuperzapper** [comment]
@dpgaspar as a christmas present I have finished this PR, can you please review it?

**Comment by megakoresh** [comment]
> @dpgaspar as a christmas present I have finished this PR, can you please review it?

Can you pls add that line to the azure provider code that I requested? Otherwise this PR will not work with azure as an oauth provider.

https://github.com/dpgaspar/Flask-AppBuilder/blob/a76ae50ab6eb019814388d3cdccb5638bee56f8b/flask_appbuilder/security/manager.py#L602

```python
return {
                "name": me["name"],
                "email": me["upn"],
                "first_name": me["given_name"],
                "last_name": me["family_name"],
                "id": me["oid"],
                "username": me["oid"],
                "groups": me.get("groups", [])
            }
```

I can of course open separate PR for this, but it seems like a bit of a waste for this one line

**Comment by jonathonbattista** [comment]
Any update on this?

**Comment by dpgaspar** [comment]
@ thesuperzapper is this ready for another review?

**Comment by thesuperzapper** [comment]
@dpgaspar yes, please

**Comment by sandhuja** [comment]
Any updates?

**Comment by thesuperzapper** [comment]
@dpgaspar FYI, the only failing check is a code coverage one (it dropped below 72.75%), but I am not sure its correctly detecting the coverage.

Regardless, I would love to hear your thoughts on the changes.

**Comment by thesuperzapper** [comment]
@dpgaspar do you know when you will get a chance to review this? 

This PR is pretty important for Airflow 2.0, given we now exclusively use FAB.

**Comment by dpgaspar** [comment]
@thesuperzapper awesome work, this looks pretty good! going to take it for a test run today, if all good I think it should be ready to merge

**Review by dpgaspar — APPROVED** [review]

File: flask_appbuilder/security/manager.py
Comment (dpgaspar) [inline flask_appbuilder/security/manager.py:904]: nit: missing annotation on `user_attributes`

File: flask_appbuilder/security/manager.py
Comment (dpgaspar) [inline flask_appbuilder/security/manager.py:1153]: we should add `f"{username}@email.notfound" or similar, if an LDAP server does not have their email fields populated, it's impossible to register multiple users this way since it hits the email DB unique constraint

**Review by dpgaspar — CHANGES_REQUESTED** [review]
After testing found the email field user registration problem

File: flask_appbuilder/security/manager.py
Comment (dpgaspar) [reply flask_appbuilder/security/manager.py:1153]: @thesuperzapper ^^

Comment (thesuperzapper) [reply flask_appbuilder/security/manager.py]: done

Comment (thesuperzapper) [reply flask_appbuilder/security/manager.py]: done @dpgaspar

File: flask_appbuilder/security/manager.py
Comment (thesuperzapper) [reply flask_appbuilder/security/manager.py]: done

File: flask_appbuilder/security/manager.py
Comment (thesuperzapper) [reply flask_appbuilder/security/manager.py]: done

File: flask_appbuilder/security/manager.py
Comment (thesuperzapper) [reply flask_appbuilder/security/manager.py]: done

File: flask_appbuilder/security/manager.py
Comment (thesuperzapper) [reply flask_appbuilder/security/manager.py]: done

**Comment by thesuperzapper** [comment]
@dpgaspar should be ready to review again

**Comment by jonathonbattista** [comment]
Any update on this?

**Comment by thesuperzapper** [comment]
@dpgaspar anything more you want to change before we merge?

**Review by dpgaspar — APPROVED** [review]
LGTM

**Comment by msardana94** [comment]
@dpgaspar Do you know when will this feature be released? This PR is super helpful and will make ldap based rbac so much better 🥇

**Comment by dpgaspar** [comment]
@msardana94  Planning to release next week

**Comment by Asturias-sam** [comment]
@dpgaspar @thesuperzapper  Is multiple role supported for `AUTH_USER_REGISTRATION_ROLE` i.e can assign role `A` and `B` both 
After this We can have multiple role using LDAP group

**Comment by Asturias-sam** [comment]
And one more thing ```
AUTH_ROLES_MAPPING = {
    "cn=fab_users,ou=groups,dc=example,dc=com &&(cn=fab_users_2,ou=groups,dc=example,dc=com)": ["User"],
    "cn=fab_admins,ou=groups,dc=example,dc=com": ["Admin"],
}
```
Is  OR AND condition supported ?

**Comment by basanthcse** [comment]
@dpgaspar @thesuperzapper 
Airflow Version 1.10.12
Flask-AppBuilder 2.3.4
We have implemented LDAP authentication with RBAC in Airflow user authentication. I have one situation where every user is able to login as Admin, if i mention AUTH_USER_REGISTRATION_ROLE = "Admin" in webserver_config.py file. I have two ldap group (airflow_admin and airflow_dev groups), i need airflow_admin user to login as ADMIN role and airflow_dev group users as OP role. How can I dynamically assign the AUTH_USER_REGISTRATION_ROLE based on the user's LDAP role?

is this feature will be include in latest version in flask-appbuilder  ?

**Comment by basanthcse** [comment]
> And one more thing
> 
> ```
> AUTH_ROLES_MAPPING = {
>    "cn=fab_users,ou=groups,dc=example,dc=com &&(cn=fab_users_2,ou=groups,dc=example,dc=com)": ["User"],
>    "cn=fab_admins,ou=groups,dc=example,dc=com": ["Admin"],
> }
> ```
> 
> Is OR AND condition supported ?

#is AUTH_ROLES_MAPPING working for you to authenticate different users from LDAP groups ?
which version are you using for airflow and flask-appbuilder ?

**Comment by thesuperzapper** [comment]
@Asturias-sam currently, we do not support OR/AND semantics in `AUTH_ROLES_MAPPING`. Note, the keys are not LDAP search terms, but the literal UID's  of groups found in the user's `AUTH_LDAP_GROUP_FIELD` after a lookup.

But most people can achieve the same outcome by creating nested groups in your LDAP/AD itself, however, I can see a need for this in OAUTH (where making groups may be harder). 

We could allow python functions as the keys of `AUTH_ROLES_MAPPING`, something like:
```python
def is_fab_user(user_role_keys: Set[str]) -> bool:
  is_users_1 = "cn=fab_users_1,ou=groups,dc=example,dc=com" in user_role_keys
  is_users_2 = "cn=fab_users_2,ou=groups,dc=example,dc=com" in user_role_keys
  return is_users_1 and is_users_2

AUTH_ROLES_MAPPING = {
    is_fab_user: ["User"],
    "cn=Admin,ou=groups,dc=example,dc=com": ["Admin"],
}
```

@Asturias-sam if you want this feature, please raise and issue, and tag me in it (probably link to this comment also)

**Pull request merged**
